### ISSUE TITLE
Design: how to make the coupling between ETM and CTM as transparant as possible

### ISSUE THREAD

**Issue by noracato** [quintel/etmodel#3899]
### Transparancy
It's not only the technical side of the coupling that is important (see  #3898). The ETM sees transparancy as one of its core values. By being transparant, users always know what is happening, can understand and check our modeling and data, and by that can trust the ETM. Being transparant is not easy, things should always be presented in an understandable and accessible way, and not get too complex or like a scary-looking wall of information.

Designing the coupling (from the ETM side) to have this kind of transparancy requires still a lot of thinking and decision making. How do we for instance show how and why certain charts in the ETM have changed, or where they now get (part of) their information from?

### Mockups
I'll include the mockups from  #3898 once more, because they could be a nice starting point for a discussion.

<div dir="row">
<a target="_blank" rel="noopener noreferrer" href="https://user-images.githubusercontent.com/14875123/150978604-451d84b3-264b-46bd-87e7-f2021d288399.png"><img src="https://user-images.githubusercontent.com/14875123/150978604-451d84b3-264b-46bd-87e7-f2021d288399.png" alt="mockup_1" style="max-width: 47%;"></a>
<a target="_blank" rel="noopener noreferrer" href="https://user-images.githubusercontent.com/14875123/150978619-efc71383-f85c-435a-8058-37d29e53a65b.png"><img src="https://user-images.githubusercontent.com/14875123/150978619-efc71383-f85c-435a-8058-37d29e53a65b.png" alt="mockup_1" style="max-width: 47%;"></a>
</div>

### Pro's and cons
| Pro's | Cons |
| --- | --- |
| + Makes the coupling more user friendly (removing one of the possible cons of  #3898) | - Might take time to come up with a right design for all the 'pieces' in the ETM (such as charts) |
| + If done right, adds a lot of extra insights | - For certain designs there could be requested changes to the CTM API |
| + Underlines one of the core values of the ETM | |
| + Shows off what the CTM can add to 'your' ETM scenario | |
| + After the design phase, individual pieces can be picked up and implemented. This allows for simultaneous development and creating smaller milestones, or an MVP and nice-to-haves | |

### Research
Before we can start implementing a transparant coupling, we first need the technical side done. However, it is wise to already start with thinking about how the coupling should be made clear to the user, as this might impact the technical design.

For creating the design, there's many questions to be answered. Here are a few proposed by @ChaelKruip in his comment on issue  #3898, that are fitting here, but please add more:
- What does a user see if they open an ETM scenario that s coupled to a CTM scenario? Should we have a "log" of all the changes made to CTM inputs and the resulting ETM energy flows?
- What will be the impact on charts and queries in the ETM if (parts of!) energy flows are set by inputs through the API without corresponding sliders?
- What will be the impact on charts and queries in the ETM if (parts of!) energy flows are set by a coupling node? Should that node feature in some of our charts as a series with a legend item?

**Time estimate for design and specifications** 3 days

**Comment by antw** [comment]
### Comparing scenarios

I really like the first mock-up. I imagine it will be useful to preview how importing a CTM scenario will affect your ETM scenario.

However, I wonder how common it will be for a user to import a CTM scenario into an *existing* ETM scenario, as opposed to creating a *new* ETM scenario with their CTM data?

I think the ability to compare any two ETM scenarios could be useful, especially if we integrate with other models in the future. If we went ahead with this I'd like to see it done in a way that isn't specific to the CTM integration. Instead, we'd create this as a generic feature, and re-use it on the "Import from CTM" page.

The ability to show slider positions doesn't depend on any existing ETM code, since we don't actually have to show real sliders; we only need to show something that *looks* like a slider. However, comparing charts would be really tricky today. They depend on so much of the ETM code – which assumes only a single scenario can be loaded at a time. We'd have to resort to iframes. I think we'd be better off waiting until we have refreshed charts that don't have this limitation. Comparing slider values would still be useful in the meantime.

### Slider positions

To show how a slider would move when importing a CTM scenario, we need to be able to calculate a new start value for the slider. Today we calculate start values with the `start_value_gql` attribute in the input file. For example:

```
- start_value_gql = present:PRODUCT(V(some_node, some_value), 100)
```

We'd need to reconsider how this works. The new starting values would have to be based on the future graph instead of the present, which I expect means getting rid of the "present:" prefix. We'd then have to work on the assumption that "present:" is implied in the GQL *except* when calculating a new start value for a disabled slider.

This also means we need to be able to differentiate between a slider that is disabled because of the CTM integration, and a slider disabled for other reasons (not supported in the dataset or a GQL calculation error).

Doing this would also allow us to show the real value of disabled sliders in the ETM – the value that results from the settings applied by the CTM – which would be a nice boost to transparency.

However, this assumes that demand set on a coupling node still flows through the rest of the graph as normal. Take this example:

![Screenshot 2022-01-31 at 14 11 45](https://user-images.githubusercontent.com/4383/151808781-2edb4e95-79c3-4e00-aacb-f7606cd54e45.png)

Imagine we have inputs which set the demand of various technologies and we add a coupling node which instead sets the total demand of *all* the technologies, rather than individual ones. If we do this, everything is fine:

![Screenshot 2022-01-31 at 14 11 48](https://user-images.githubusercontent.com/4383/151808802-2571f945-e817-407f-9fd9-ab862742281c.png)

With this setup, demand will flow from left-to-right, and we'll be able to calclulate the demand of each individual technology, and show its value next to a slider in the ETM.

On the other hand, if the coupling node demand "skips" the technologies, we will *not* be able to calculate any demand for the technologies. Presumably their demand would have to be set to zero:

![Screenshot 2022-01-31 at 14 11 50](https://user-images.githubusercontent.com/4383/151808811-4d6a831c-dd08-4869-bb50-b4776e98f715.png)

**Comment by antw** [comment]
> On the other hand, if the coupling node demand "skips" the technologies, we will not be able to calculate any demand for the technologies.

I think this is the way the proposed design works. In https://github.com/quintel/etengine/issues/1222#issuecomment-1026733809 the coupling node skips over a number of nodes. We won't be able to calculate meaningful input values for any nodes or edges which are skipped by the coupling.

This isn't a show-stopper, but it may limit our ability to provide a meaningful comparison of slider values for any of the affected nodes and edges.

---
Including @mabijkerk.

**Comment by mabijkerk** [comment]
Thanks for linking me to this discussion. The proposed design for the fertiliser sector is not yet definitive; we can still make changes. My motivation for skipping the technology nodes is because it requires taking the burner losses into account. This would make the input statements for final demands per carrier more complicated.

More complicated input statements aren't a showstopper either, and a quick solution may be to update the losses of the technology to 0% when the coupling node is 'active'.

**Comment by mabijkerk** [comment]
A motivation for **not** passing the demand through the technologies is that a single technology coupling node makes it much easier to set the costs for an externally coupled model.
